### PR TITLE
[5.2] Add method to implement multiple scopes with variable names 

### PR DESCRIPTION
**Pull request by sileence**
Proposal to add the method `scopes` to allow the implementation of multiple variable scopes at once, for example:

`Post::scopes(['question', 'pending' => true])->get()` is the same as:

`Post::questions()->pending(true)->get()`

### CODE REVIEW

**Comment by JosephSilber** [comment]
I don't see how using an array is in any way better than calling them directly.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -1143,6 +1143,28 @@ protected function parseNestedWith($name, $results)
...
+     *
+     * @param  array  $scopes
+     * @return mixed
+     */
+    public function scopes(array $scopes)
+    {
+        $builder = $this;
+
+        foreach ($scopes as $scope => $parameters) {
+            if (is_numeric($scope)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Maybe this should be `is_int`?

**Comment by sileence** [comment]
@JosephSilber 

I have something like this in my code:

```
        $scopes = $this->getListScopes($request);

        $posts = Post::scopes($scopes)->paginate();
```

Based on the request I determine the dynamic scope to filter the posts. Doing the same with chained methods is not that easy and don't look that nice.

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 1 above)
Comment (sileence) [inline src/Illuminate/Database/Eloquent/Builder.php]: sure

**Comment by JosephSilber** [comment]
Then this is something very unique to your app, and you should keep it there.

Doesn't make much sense in the framework IMO.

**Comment by GrahamCampbell** [comment]
Hmmm, I think it could be quite useful. Could make for nicer design rather than making method calls using variables as method names. I'm :+1: here.

**Comment by taylorotwell** [comment]
Can you send a PR for a test for this?

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 2]
```diff
@@ -1143,6 +1143,28 @@ protected function parseNestedWith($name, $results)
...
+    {
+        $builder = $this;
+
+        foreach ($scopes as $scope => $parameters) {
+            if (is_int($scope)) {
+                $scope = $parameters;
+                $parameters = [];
+            }
+
+            $builder = $builder->callScope('scope'.ucfirst($scope), (array) $parameters);
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: @sileence This line doesn't work. ``callScope`` needs ``callable`` type as the first argument, so it should be:

```php
$callable = ['scope'.ucfirst($scope), $this->model];

$builder->callScope($callable, (array) $parameters)
```

Also, scope can return anything, not necessarily a builder instance, so I wouldn't re-assign it here.

Comment (sileence) [inline src/Illuminate/Database/Eloquent/Builder.php]: working on that.

Comment (sileence) [inline src/Illuminate/Database/Eloquent/Builder.php]: #14078

**Pull request merged**